src/r_main.cpp

```diff
@@ -608,52 +609,133 @@ void R_SetupFreelook()
 
 //==========================================================================
 //
-// R_EnterMirror
+// R_EnterPortal
 //
 // [RH] Draw the reflection inside a mirror
+// [ZZ] Merged with portal code, originally called R_EnterMirror
 //
 //==========================================================================
 
-void R_EnterMirror (drawseg_t *ds, int depth)
+CVAR(Int, r_portal_recursions, 4, CVAR_ARCHIVE)
+CVAR(Bool, r_highlight_portals, false, CVAR_ARCHIVE)
+
+void R_HighlightPortal (PortalDrawseg* pds)
 {
+	// [ZZ] NO OVERFLOW CHECKS HERE
+	//      I believe it won't break. if it does, blame me. :(
+
+	BYTE color = (BYTE)BestColor((DWORD *)GPalette.BaseColors, 255, 0, 0, 0, 255);
+
+	BYTE* pixels = RenderTarget->GetBuffer();
+	// top edge
+	for (int x = pds->x1; x <= pds->x2; x++)
+	{
+		if (x < 0 || x >= RenderTarget->GetWidth())
+			continue;
+
+		int p = x - pds->x1;
+		int Ytop = pds->ceilingclip[p];
+		int Ybottom = pds->floorclip[p];
+
+		if (x == pds->x1 || x == pds->x2)
+		{
+			RenderTarget->DrawLine(x, Ytop, x, Ybottom+1, color, 0);
+			continue;
+		}
+
+		int YtopPrev = pds->ceilingclip[p-1];
+		int YbottomPrev = pds->floorclip[p-1];
+
+		if (abs(Ytop-YtopPrev) > 1)
+			RenderTarget->DrawLine(x, YtopPrev, x, Ytop, color, 0);
+		else *(pixels + Ytop * RenderTarget->GetPitch() + x) = color;
+
+		if (abs(Ybottom-YbottomPrev) > 1)
+			RenderTarget->DrawLine(x, YbottomPrev, x, Ybottom, color, 0);
+		else *(pixels + Ybottom * RenderTarget->GetPitch() + x) = color;
+	}
+}
+
+void R_EnterPortal (PortalDrawseg* pds, int depth)
+{
+	// [ZZ] check depth. fill portal with black if it's exceeding the visual recursion limit, and continue like nothing happened.
+	if (depth > r_portal_recursions)
+	{
+		BYTE color = (BYTE)BestColor((DWORD *)GPalette.BaseColors, 0, 0, 0, 0, 255);

```

Comment:
The code in R_HighlightPortal draws red lines around the drawseg. The code in R_EnterPortal fills the whole drawseg area with black color.